File: 1054. 求平均值.py
import sys
def islegal( num ) :
    flag = True
    dotcnt = 0
    num = num[ 1 : ] if num[ 0 ] == '-' or num[ 0 ] == '+' else num
    for index , i in enumerate( num ) :
        if ( not i.isdigit() ) and i != '.' :
            flag = False
            break
        elif i == '.' :
            dotcnt += 1
            if len( num[ index + 1 : ] ) > 2 or dotcnt > 1 :
                flag = False
                break
    if flag :
        num = float( num )
        if -1000 > num or num > 1000 :
            flag = False
    return flag
    
N = int( sys.stdin.readline() )
Sum = 0
cnt = 0
for i in sys.stdin.readline().split() :
    if islegal( i ) :
        i = float( i )
        Sum += i
        cnt += 1
    else :
        print( 'ERROR: %s is not a legal number' %i )
if not cnt :
    print( 'The average of 0 numbers is Undefined' )
else :
    print( 'The average of {} number{} is {:.2f}'.format( cnt ,'s' if cnt > 1 else '', Sum / cnt ) )

File: 1054. 求平均值.c
#include<stdio.h> //用到了strtof 将字符串转换为浮点数，并在不可转换的位置停止并把地址赋给end变量。
#include<stdlib.h>
#include<string.h>
#include<stdbool.h>
bool IsLegal( char [] ) ;
int main(void)
{
        int N ;
        float sum = 0 ;
        int cnt = 0 ;
        scanf( "%d" , &N ) ;
        char Data[ 1000 ] ;
        for( int i = 0 ; i < N ; i++ )
        {
                scanf( "%s" , Data ) ;
                if( IsLegal( Data ) )
                {
                        cnt++ ;
                        sum += atof( Data ) ;
                }
                else
                        printf("ERROR: %s is not a legal number\n", Data ) ;
        }
        printf(!cnt?"The average of 0 numbers is Undefined":"The average of %d number%s is %.2f",cnt,cnt==1?"":"s",sum/cnt);
        return 0 ;
}
bool IsLegal( char Data[] )
{
        bool flag = true ;
        int Length = strlen( Data ) ;
        char *end ;
        float num = strtof( Data , &end ) ;
        if( *end )//如果end不为空字符 说明遇到了不能转换的字符。则该字符串不合法
        {
                flag = false ;
        }
        if( flag )//判断小数点后位数是否超过2位
        {
                int index = Length - 1 ;
                for( int i = 0 ; i < Length ; i++ )
                        if( Data[ i ] == '.' )
                        {
                                index = i ;
                                break ;
                        }
                if( Length - index - 1 > 2 )
                        flag = false ;
        }
        if( flag )//字符串为数字，判断范围是否合法
                if( num > 1000 || num < -1000 )
                        flag = false ;
        return flag ;
}
